File: lib/src/web_socket/browser/sockjs.dart
// Copyright 2015 Workiva Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'dart:async';
import 'dart:js' as js;

import 'package:w_transport/src/web_socket/browser/sockjs_port.dart';
import 'package:w_transport/src/web_socket/browser/sockjs_wrapper.dart';
import 'package:w_transport/src/web_socket/common/web_socket.dart';
import 'package:w_transport/src/web_socket/web_socket.dart';

/// Implementation of the platform-dependent pieces of the [WebSocket] class for
/// the SockJS browser configuration. This class uses the SockJS library to
/// establish a WebSocket-like connection (could be a native WebSocket, could
/// be XHR-streaming).
abstract class SockJSWebSocket extends CommonWebSocket implements WebSocket {
  static Future<WebSocket> connect(Uri uri,
      {bool debug: false,
      bool noCredentials: false,
      List<String> protocolsWhitelist,
      Duration timeout}) async {
    // The SockJS wrapper library is preferred because it uses the actual JS lib
    // which is community-supported and fully-featured. But, it requires that
    // the sockjs.js file is included. We can check for that by checking for the
    // existence of a `SockJS` object on the window.
    if (js.context.hasProperty('SockJS')) {
      return SockJSWrapperWebSocket.connect(uri,
          debug: debug,
          noCredentials: noCredentials,
          protocolsWhitelist: protocolsWhitelist,
          timeout: timeout);
    }

    // If the sockjs.js file wasn't detected, then we fallback to the original
    // SockJS Dart port.
    return SockJSPortWebSocket.connect(uri,
        debug: debug,
        noCredentials: noCredentials,
        protocolsWhitelist: protocolsWhitelist,
        timeout: timeout);
  }
}

File: tool/sockjs.js
// Copyright 2015 Workiva Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

try {
    var http = require('http');
    var sockjs = require('sockjs');
} catch (e) {
    console.error('Missing NPM dependencies: "http" and "sockjs".\nRun `npm install`.');
    process.exit(1);
}


var echo = sockjs.createServer();
echo.on('connection', function(conn) {
    conn.on('data', function(message) {
        console.log('[echo] "' + message + '"');
        conn.write(message);
    });
    conn.on('close', function() {
        console.log('[echo] closed.');
    });
});

var closeOnRequest = sockjs.createServer();
closeOnRequest.on('connection', function(conn) {
    conn.on('data', function(message) {
        if (message.substr(0, 'close'.length) === 'close') {
            var parts = message.split(':');
            var code;
            var reason;
            if (parts.length >= 2) {
                code = parseInt(parts[1], 10);
            }
            if (parts.length >= 3) {
                reason = parts[2];
            }
            console.log('[cor] close request received');
            conn.close(code, reason);
        }
    });
    conn.on('close', function() {
        console.log('[cor] closed.');
    });
});

var ping = sockjs.createServer();
ping.on('connection', function(conn) {
    conn.on('data', function(message) {
        message = message.replace('ping', '');
        var numPongs = 1;
        if (message.length > 0) {
            try {
                numPongs = parseInt(message, 10);
            } catch (e) {}
        }
        for (var i = 0; i < numPongs; i++) {
            conn.write('pong');
            console.log('[ping] pong');
        }
    });
    conn.on('close', function() {
        console.log('[ping] closed.');
    });
});

var server = http.createServer();
closeOnRequest.installHandlers(server, {prefix: '/test/ws/close'});
echo.installHandlers(server, {prefix: '/test/ws/echo'});
echo.installHandlers(server, {prefix: '/example/ws/echo'});
ping.installHandlers(server, {prefix: '/test/ws/ping'});
server.listen(8026, 'localhost');
